### ISSUE TITLE
Stuck on Builder: Calling PyInstaller. Please wait...

### ISSUE THREAD

**Issue by YadaFada** [sithis993/Crypter#54]
Hi, 
First of all, Amazing job on this builder!

Everything was going fine, except for one part, when i started building, the builder stays on "Builder: Calling PyInstaller. Please wait...", i tried waiting for over an hour but no progress happened. I would really appreciate the help :(
Here's the full log : 
[2018-04-10 11:05:06]: Build Launched
[2018-04-10 11:05:06]: DEBUG Level: 1 - Low
[2018-04-10 11:05:06]: Builder: Checking configuration...
[2018-04-10 11:05:06]: Builder: Checking builder_language
[2018-04-10 11:05:06]: Builder: Checking pyinstaller_aes_key
[2018-04-10 11:05:06]: Builder: Checking icon_file
[2018-04-10 11:05:06]: Builder: Checking open_gui_on_login
[2018-04-10 11:05:06]: Builder: Checking gui_title
[2018-04-10 11:05:07]: Builder: Checking upx_dir
[2018-04-10 11:05:07]: Builder: Checking delete_shadow_copies
[2018-04-10 11:05:07]: Builder: Checking disable_task_manager
[2018-04-10 11:05:07]: Builder: Checking key_destruction_time
[2018-04-10 11:05:07]: Builder: Checking wallet_address
[2018-04-10 11:05:07]: Builder: Checking bitcoin_fee
[2018-04-10 11:05:07]: Builder: Checking encrypt_attached_drives
[2018-04-10 11:05:07]: Builder: Checking encrypt_user_home
[2018-04-10 11:05:07]: Builder: Checking max_file_size_to_encrypt
[2018-04-10 11:05:08]: Builder: Checking filetypes_to_encrypt
[2018-04-10 11:05:08]: Builder: Checking encrypted_file_extension
[2018-04-10 11:05:08]: Builder: Checking make_gui_resizeable
[2018-04-10 11:05:08]: Builder: Checking always_on_top
[2018-04-10 11:05:08]: Builder: Checking background_colour
[2018-04-10 11:05:08]: Builder: Checking heading_font_colour
[2018-04-10 11:05:08]: Builder: Checking primary_font_colour
[2018-04-10 11:05:08]: Builder: Checking secondary_font_colour
[2018-04-10 11:05:08]: Builder: Checking ransom_message
[2018-04-10 11:05:08]: Builder: Checking debug_level
[2018-04-10 11:05:08]: Builder: Validation successful
[2018-04-10 11:05:09]: Builder: Encryption will target attached drives and the user's home directory
[2018-04-10 11:05:09]: Builder: Creating binary runtime config at Resources/runtime.cfg
[2018-04-10 11:05:09]: Builder: Runtime config successfully written
[2018-04-10 11:05:09]: Builder: Creating PyInstaller SPEC file
[2018-04-10 11:05:09]: Spec: PyInstaller binary Icon file specified. Custom icon will be added
[2018-04-10 11:05:09]: Builder: (Warning): UPX path not specified. The PyInstaller binary will not be packed. It is recommended that UPX is used as this can reduce the binary size by several Megabytes
[2018-04-10 11:05:09]: Builder: SPEC file successfully created
[2018-04-10 11:05:09]: Builder: Calling PyInstaller. Please wait...

Thank you so much :)

**Comment by sithis993** [comment]
Hey, thanks! :-) sorry that you're having problems.

The call to PyInstaller should be pretty much instant, so there's definitely something not quite right here. Have you definitely installed PyInstaller? If you open a cmd prompt and just run "pyinstaller", you should get an output which looks like this:

![image](https://user-images.githubusercontent.com/22727993/38549151-570c6d6c-3cab-11e8-9dd4-df9ded3068aa.png)

**Comment by YadaFada** [comment]
Well that's weird, I'm pretty sure i have Pyinstaller installed, I even had it installed from cmd, it was downloaded and installed successfully.
![1](https://user-images.githubusercontent.com/38243857/38551638-606b37a0-3cc2-11e8-9df7-0af9f48653a5.jpg)


By the way, i tried running the builder form cmd and here's what i got in the log where it got stuck on "Builder: Calling PyInstaller. Please wait..." 

![2](https://user-images.githubusercontent.com/38243857/38552112-c5c07e52-3cc3-11e8-96a6-45457bd4ec3d.jpg)

So apparently something is missing, Can you tell what it is? 
Thanks!!

**Comment by sithis993** [comment]
Yeah I've sometimes seen PyInstaller do this before, but not for a while... You might have it installed but it just isn't on the path for some reason. I've had someone experience the same issue in the past, so I think I'm going to add some handling in the future.

Anyway,  can you try installing PyInstaller from [here](https://github.com/pyinstaller/pyinstaller/releases/download/v3.1.1/PyInstaller-3.1.1.tar.gz). Just extract the contents and run _python setup.py_

Once you've done that, try closing and reopening the cmd prompt and running _pyinstaller_ again. Hopefully this time you should see the PyInstaller prompt output. If not, let me know and there are a few other things we can try :-)

**Comment by YadaFada** [comment]
Hi, 
I tried exactly what you told me, the installation was successful, but then again i got the exact same error message :( 
sorry for the bother, and thanks again for your time.

**Comment by sithis993** [comment]
Okay that's interesting. So you've definitely got PyInstaller installed, but the builder can't find it and you can't run it from a CMD prompt. That says to me that the directory where PyInstaller is installed is not on the Windows path

I see you've installed Python at C:\Python27. There should be a pyinstaller.exe in C:\Python27\Scripts. Can you take a look and see if there is? if so, then you'll need to add that directory to your Windows path so that the builder can find it :-). Do you know how to do this?

**Comment by YadaFada** [comment]
Yes, it's true, i found pyinstaller.exe in C;\Pythono27\Scripts. 
I'm not sure if what I'm following you exactly about adding the directory to Windows, its not just a copy-paste thing right ? :-D
However if its adding "C:\Python27" as a path from "Advanced windows settings" then i've already done that. 
Thank you!

**Comment by YadaFada** [comment]
Hey!
I got it working! 
Turns out yes i needed to add " C:\Python27\Scripts" too in Advanced System Settings.
Everything worked perfectly now and i got my Crypter <3
I will test it out later today :) 

Just one tiny question, I wanna know how strong this python crypter is and how far it could get. 
Say it was planted on a computer that was connected to a couple other computers that were all on the same server (so they have the exact same local disks), will the crypter be able to encrypt all the files on all the local disks?

**Comment by sithis993** [comment]
Aha, I was just about to say that and then saw your message! glad to see you've at least got the builder working correctly :-)

That depends on the environment and how they're connected really. From your description, the first thing that comes to mind is a virtual environment running on VMWare or Hyper-V, where the host machine is running several guests. In this scenario, they all have their own virutal disks, but they could be running on the same physical disk that belongs to the VMWare/Hyper-V host. Is this the kind of thing you mean?

Can you elaborate a bit on what you mean by the machines being on the same server, and having the same local disks? also how would they be connected?

**Comment by YadaFada** [comment]
The example you gave is quite close, but let me draw you a full picture:
Lets say we have a main server pc with Windows Server 2008 on it. This computer has a couple of other computers connected to it (Via Cable), each computer has its own windows, but when it comes to files, all of them share the same files and Local Disks (C: D: ...etc), But each computer can only see its own Local Disk. Lets say we have 5 computers so we have 5 local disks, So for example computer 1 has Local Disk (D:) and can only see (D:) the others are hidden. To clarify this further:
                            -->Computer 1-->Local Disk (D:) 
                            -->Computer 2-->Local Disk (E:) 
Server Computer -->Computer 3-->Local Disk (F:) 
                            -->Computer 4-->Local Disk (G:) 
                            -->Computer 5-->Local Disk (H:) 
So the Server Computer has all of the data of all the computers and they are all connected together.
Say we launched the Crypter on computer 3 for example, will this affect all the other computers on the server?

**Comment by YadaFada** [comment]
Hey, 
I just ran the test on my test computer, and after double clicking on the crypter, i wait a little then i get the key.txt, then after like 1 minute i get an error massage saying "Fatal Error!" : "Main Returned -1"  
This is so sad please tell me this is fixable :((
Thank you!

**Comment by sithis993** [comment]
Hey, sorry for the slow reply this time, I've been a little busy the last few days! I think I somewhat understand what you're saying, but you might be more suited to answer the question yourself if I explain how Crypter finds files to encrypt. 

When it's run, Crypter looks for disks that are attached to the victim machine and then searches those disks for matching files. So let's say your server computer is connected to two network shares, and also has a USB drive plugged in. Crypter will encrypt the contents of both network shares, as well as the USB drive.

To return to your example, if the server computer in this case has the local disks of all of the others computers mounted to it as file shares, then yes they will be encrypted. However, they have to be mounted as shares so that to the server these disks are attached as H: F: G: etc.

Does that make sense?

**Comment by YadaFada** [comment]
Hi! 
Thank you for your reply, no worries about being late :) 
So i did rebuild the crypter with the new version, but for some reason its searching for pyinstaller.exe in C:\python27 instead of C:\python27\Scripts  . See this : 
[2018-04-14 03:27:56]: Build Launched
[2018-04-14 03:27:56]: DEBUG Level: 1 - Low
[2018-04-14 03:27:56]: Builder: Checking configuration...
[2018-04-14 03:27:56]: Builder: Checking builder_language
[2018-04-14 03:27:56]: Builder: Checking pyinstaller_aes_key
[2018-04-14 03:27:56]: Builder: Checking icon_file
[2018-04-14 03:27:57]: Builder: Checking open_gui_on_login
[2018-04-14 03:27:57]: Builder: Checking gui_title
[2018-04-14 03:27:57]: Builder: Checking upx_dir
[2018-04-14 03:27:57]: Builder: Checking delete_shadow_copies
[2018-04-14 03:27:57]: Builder: Checking disable_task_manager
[2018-04-14 03:27:57]: Builder: Checking key_destruction_time
[2018-04-14 03:27:57]: Builder: Checking wallet_address
[2018-04-14 03:27:57]: Builder: Checking bitcoin_fee
[2018-04-14 03:27:57]: Builder: Checking encrypt_attached_drives
[2018-04-14 03:27:58]: Builder: Checking encrypt_user_home
[2018-04-14 03:27:58]: Builder: Checking max_file_size_to_encrypt
[2018-04-14 03:27:58]: Builder: Checking filetypes_to_encrypt
[2018-04-14 03:27:58]: Builder: Checking encrypted_file_extension
[2018-04-14 03:27:58]: Builder: Checking make_gui_resizeable
[2018-04-14 03:27:58]: Builder: Checking always_on_top
[2018-04-14 03:27:58]: Builder: Checking background_colour
[2018-04-14 03:27:58]: Builder: Checking heading_font_colour
[2018-04-14 03:27:58]: Builder: Checking primary_font_colour
[2018-04-14 03:27:59]: Builder: Checking secondary_font_colour
[2018-04-14 03:27:59]: Builder: Checking ransom_message
[2018-04-14 03:27:59]: Builder: Checking debug_level
[2018-04-14 03:27:59]: Builder: Validation successful
[2018-04-14 03:27:59]: Builder: Encryption will target attached drives and the user's home directory
[2018-04-14 03:27:59]: Builder: Creating binary runtime config at Resources/runtime.cfg
[2018-04-14 03:27:59]: Builder: Runtime config successfully written
[2018-04-14 03:27:59]: Builder: Creating PyInstaller SPEC file
[2018-04-14 03:27:59]: Spec: PyInstaller binary Icon file specified. Custom icon will be added
[2018-04-14 03:27:59]: Builder: (Warning): UPX path not specified. The PyInstaller binary will not be packed. It is recommended that UPX is used as this can reduce the binary size by several Megabytes
[2018-04-14 03:27:59]: Builder: SPEC file successfully created
[2018-04-14 03:27:59]: Builder: Calling PyInstaller. Please wait...
[2018-04-14 03:27:59]: (ERROR): Builder: Could not find PyInstaller at 'C:\Python27\pyinstaller.exe'. Check that PyInstaller is installed
[2018-04-14 03:27:59]: Build finished with error


I tried my luck by copying the 3 pyinstaller  files in Scripts to C:\python27    XD
it ended up building correctly but once i launched the crypter i got an error saying that the file is corrupt :D, so how can i change the reading path ? 


Oh, and thanks for the information up there, yes i do understand it now :)

**Comment by sithis993** [comment]
I'm going to assume this is the error you're seeing. The guy who opened issue #53 also asked the same question, so I'm going to paste my response from there and hopefully you can understand what's going on :-). Here it is:

I suspect it is. This behaviour is actually completely intentional :-) This should mean you have it working! Let me explain what happens. 

Every time Crypter.exe runs it creates a Mutex to make sure that only _one_ instance can run at any one time. This is so that whilst Crypter is busy encrypting files, the user can't open the exe again and have two (or more) copies encrypting files simultaneously. If this did happen, then the user could run into all kinds of problems, which would ultimately mean that they couldn't get their files back. Make sense?

Now onto the error message: If Crypter.exe runs and there is already another instance running, then I purposely display this error message. When I added this functionality, I had a social engineering use case in mind. 

Let's say you send an Email to someone with Crypter.exe attached. They open it and run the exe, but from what they can see, nothing is happening. What they don't know is that in the background the encrypting operation is encrypting all of the files on the machine. What do you think the user will do first? they will try to open the exe again. Now if I didn't have the Mutex, another instance of Crypter would begin encrypting files, which is why I added the Mutex functionality. This error message is there so that a user would see it and assume that the file was broken and carry on with their day. Thus, there isn't really an error... It was done intentionally to minimise suspicion and increase the likelihood of the encryption operation finishing without the user interrupting it.

So if you're getting this, don't worry. Click OK, then open Task Manager and see if there is already a Crypter.exe process running :-)

**Comment by YadaFada** [comment]
Oh my god this is absolutely true! I just re opened my test computer to find crypter GUI in my face xD 
its working!! so back then when i launched it, it was working!! You are truly the best hahaha thank you :) 
But Im curious as to why the GUI didnt appear back then when i opened it first? is it because it appears after encrypting everything? which would be highly likely because my test computer is SUPER slow. 
last but not least, say I closed the GUI window from the X, the only way to get it back would be to restart my computer ? 
I'm so grateful to you seriously thank you :)

**Comment by sithis993** [comment]
Haha nice! thanks, no problem :-). Thank you as well! Most of the time people raise issues but then don't help at all in trying to resolve them. I usually just get "help, it's not working, please fix it", so I really appreciate your patience and assistance in getting this fixed for you, and for others.

Yep the GUI doesn't appear until the encryption has finished. You wouldn't want the user to know what's going on until it's finished right? ;-) Starting the encryption _after_ the GUI appears would provide an opportunity to backup files which haven't yet been encrypted, disconnect file shares etc.

Nope, if you close the GUI, you can simply re-run the .exe. When Crypter starts up it checks to see if it has already executed and encrypted files. If it hasn't then it starts the encryption, but if it has then it just opens the GUI :-).

**Comment by YadaFada** [comment]
Brilliant! I can't thank you enough for this, you're awesome :) 
This has been really fun and useful!
I wish you the best of luck ahead of you.
Cheers!
